### ISSUE TITLE
R8-D — requalify Qwen3.8 UD-IQ1_S NVIDIA RPC under validated true-greedy decoding

### ISSUE THREAD

**Issue by ezutfen** [Zutfen-LLC/inferswarm#195]
## Status

**AGENT-READY — PHYSICAL REQUALIFICATION / TRUE-GREEDY / NVIDIA-FIRST / TEXT-ONLY / SINGLE-SLOT**

Parent: #188 — R8 Qwen3.8-Flash-Next heterogeneous residency/execution program.

Accepted predecessor: #193 / PR #194, merged as `f65b70970a9a10bf57bd58a902fe6e08b588c619`.

Accepted R8-C terminal:

`R8C_QWEN38_RPC_DIVERGENCE_CAUSE_LOCALIZED`

Start from current `origin/main` at/after `f65b70970a9a10bf57bd58a902fe6e08b588c619`. If main advances before execution, reconcile ordinary documentation/test drift normally. Any execution-bearing change to the R8 runtime/model/topology path requires an explicit applicability audit before physical authority is frozen.

This issue performs the narrowly recommended R8-C successor: rerun the R8-B NVIDIA cross-host RPC qualification under a **sampler contract that is mechanically proven to be true greedy before candidate output**. It does not repair llama.cpp, change the model/quantization, add AMD/Vulkan, change InferSwarm ordinary serving, or optimize model-source locality.

---

## Accepted predecessor finding — consume, do not reinterpret

R8-C established that the accepted R8-B FAIL was **not evidence of RPC transport/state corruption**.

At pinned llama.cpp `b29c606e28a01b1bc8c1351026a0fa6e616bf6c4`:

- the R8-B request used `samplers=["greedy"]`;
- `greedy` did not resolve;
- the accepted reference log retained 16 `unable to match sampler by name 'greedy'` warnings;
- the actually built chain was `logits -> dist`;
- both arms therefore executed seeded distribution sampling (`seed=0`), not argmax greedy;
- lawful cross-backend numerical-path differences were amplified into token divergence;
- matched loopback-vs-remote RPC was byte-identical, exonerating the cross-host transport path for the reproduced wedge;
- a request-only intervention using `samplers=["top_k"]`, `top_k=1` materially improved agreement: exact equality on the 1024/3072 diagnostic cases and a longer common prefix on the short case;
- residual true-greedy differences remained on the short case after position 5 and on the long-context case, so no PASS is presumed.

The accepted R8-B FAIL remains immutable and must not be rewritten or retroactively converted to PASS.

---

## Frozen subject — preserve unless a pre-output contradiction forces BLOCKED

Use the accepted R8-B/R8-C subject unchanged:

- official model authority: `Qwen/Qwen3.8-Flash-Next@de4b8e4d43b917e7706784d8bb445c9af86a3540`;
- GGUF conversion authority: `unsloth/Qwen3.8-Flash-Next-GGUF@38bb39ee97821de2c9009abb7e93950eec396e66`;
- representation: `UD-IQ1_S`;
- exact three-member split set and accepted LFS SHA-256 identities from R8-A;
- total encoded bytes: `72,546,461,344`;
- runtime: llama.cpp `b29c606e28a01b1bc8c1351026a0fa6e616bf6c4` (`v0.4.1`), exact accepted binary hashes unless an explicit pre-output applicability audit proves a byte-identical rebuilt artifact from the same source/toolchain;
- text-only, no MTP/speculative, no multimodal;
- accepted host-side lazy PLE/n-gram placement semantics;
- accepted NVIDIA candidate topology spanning inferswarm01 + inferswarm03 + inferswarm04, five CUDA devices total;
- accepted non-RPC reference shape: same runtime/model bytes, single-host path using local CUDA + CPU spill;
- accepted R8-B exact prompt-token fixtures and lengths: 256 / 1022 / 3077 / 4097 tokens;
- `n_predict=8` and the accepted stop policy.

Re-measure current mutable host/GPU/BDF/UUID/driver/runtime/link identities before output. Historical inventory is evidence, not a permanent assumption.

Do **not** change the model-source/backing transport in this issue merely to reduce startup time. The local-verified-model-backing optimization is a separate successor so this requalification changes the sampler contract, not the artifact source path, at the same time.

---

## Objective

Produce one clean physical requalification answering:

> With the exact accepted Qwen3.8 UD-IQ1_S bytes, pinned llama.cpp runtime and accepted NVIDIA reference/candidate topology, does the cross-host RPC candidate satisfy the original R8-B correctness/restart/residency contract when both arms are prospectively proven to execute **true greedy argmax decoding**?

Performance victory is not required.

---

## Phase 0 — reconcile main and freeze R8-D authority

Before correctness-bearing output:

1. record exact `origin/main` and prove merge `f65b70970a9a10bf57bd58a902fe6e08b588c619` ancestral;
2. re-verify accepted R8-A, R8-B and R8-C manifests/bundles byte-unchanged;
3. re-hash the exact llama.cpp binaries on every selected host;
4. re-hash all three accepted GGUF split members against R8-A authority;
5. retain fresh raw host/GPU/NIC/runtime inventory for every participating node;
6. freeze exact reference and candidate commands/process roles/ports/topology;
7. freeze exact request JSON and sampler configuration;
8. freeze exact four-case prompt token IDs, `n_predict=8`, stop semantics and comparison rule;
9. create a fresh additive evidence namespace, suggested `docs/investigations/qwen38-flash-next-r8-d/`;
10. create a fresh campaign/authorization identity and bind every correctness-bearing producer/hash before model output;
11. freeze launch order, repeat counts, restart control, STOP conditions and terminals;
12. commit/push all correctness-bearing harness/authority bytes before physical candidate output.

After authority freeze, correctness-bearing source/config bytes are immutable for the campaign. Any required change supersedes the affected attempt and requires a fresh freeze before new output.

---

## Phase 1 — mechanically prove the true-greedy sampler contract

The defect that invalidated the intended R8-B greedy protocol must be impossible to repeat silently.

Canonical R8-D request contract:

```json
{
  "samplers": ["top_k"],
  "top_k": 1,
  "temperature": 0.0,
  "seed": 0,
  "cache_prompt": false,
  "stream": false,
  "return_tokens": true
}
```

Use the exact accepted prompt token IDs and add `n_predict=8` per case.

Before distributed candidate output, prove mechanically:

1. `top_k` resolves on the pinned runtime; no `unable to match sampler` warning is present;
2. runtime/source inspection demonstrates `top_k=1` is applied before the terminal draw, leaving exactly one admissible token;
3. the observed built sampler chain is compatible with that contract (for example `logits -> top-k -> dist`; terminal `dist` is acceptable only because `top_k=1` has reduced support to one token);
4. no default/implicit sampler silently widens the support after top-k;
5. request parsing preserves `top_k=1` exactly;
6. a negative control with the old `samplers=["greedy"]` is rejected/classified as noncanonical and reproduces the unresolved-name warning without being accepted as R8-D output;
7. a negative control with `top_k>1` cannot be mislabeled `true_greedy`;
8. a missing/unknown sampler-chain observation is BLOCKED, not assumed correct.

Do **not** use the pinned server's `n_probs` rows as the correctness oracle: R8-C demonstrated row misalignment. No margin claim may be derived from that surface.

If an independent argmax observation is needed to validate the sampler contract, use a bounded observation path that is proven non-perturbing before relying on it. Do not repair the runtime in this issue.

If true-greedy execution cannot be prospectively proven on the pinned build, STOP `R8D_EVIDENCE_BLOCKED` before distributed correctness output.

---

## Phase 2 — freeze a fresh true-greedy non-RPC reference

Using the exact accepted R8-B non-RPC reference placement and the now-proven true-greedy request:

1. run the exact four accepted fixtures: 256 / 1022 / 3077 / 4097 prompt tokens;
2. generate up to 8 committed tokens per case under the frozen stop policy;
3. run a minimum 3 clean repeats per case;
4. require byte-identical token IDs and stop semantics across repeats before freezing the reference;
5. retain raw request/response bytes, server logs, process/device identities, placement/load evidence and sampler-contract evidence;
6. freeze the resulting reference outputs/digests **before the first distributed candidate output**.

If the reference is not deterministic under true greedy, STOP `R8D_EVIDENCE_BLOCKED`; do not create an after-the-fact comparator.

---

## Phase 3 — canonical distributed true-greedy RPC qualification

Run the exact same frozen fixtures/request contract through the accepted five-CUDA-device cross-host candidate.

For each case require:

- exact pinned runtime/model/split identities;
- exact intended process/GPU/BDF/UUID set;
- accepted PLE/n-gram placement semantics;
- intended layer/tensor placement;
- no silent CPU-only/wrong-device/alternate-topology fallback;
- mechanically proven true-greedy sampler contract on the candidate launch;
- clean completion with no NaN/Inf/pathological all-zero output;
- exact generated token-ID equality to the frozen true-greedy reference;
- exact committed count and stop semantics;
- exact request/fixture identity.

A valid mismatch is a qualification FAIL. Do not weaken the comparator, change the sampler, change topology, change runtime, or selectively rerun a differing case seeking a pass.

Record first divergence position for diagnostics only.

---

## Phase 4 — repeatability and clean restart

For at least one short sentinel and one >2K sentinel:

1. run at least 3 candidate repeats and require candidate-local determinism;
2. terminate exact client/RPC processes;
3. prove old process identities are gone;
4. relaunch the same pinned binaries/configuration;
5. mechanically re-prove the true-greedy sampler contract after restart;
6. rerun the sentinels;
7. require exact equality to the pre-restart candidate outputs and, for a PASS campaign, to the frozen reference.

This is a process restart, not node reboot/failover.

---

## Phase 5 — residency/resource/transport accounting

Reuse the R8-B accounting contract and retain current evidence for:

- GGUF backing/mmap/file-backed mappings;
- host RSS/PSS where available;
- PLE/n-gram host placement;
- device-local model/weight bytes;
- KV/GDN/QSA/recurrent/session state;
- compute/workspace and staging buffers;
- RPC server/client memory;
- GPU memory before load / ready / during cases;
- initialization model-weight transfer bytes/time where observable;
- steady-state network bytes/time;
- unexpected persistent anonymous host mirrors;
- wrong-device/fallback/rematerialization/source anomalies.

Do not reinterpret a full file-backed mmap as an anonymous persistent host mirror.

This phase is evidence/accounting only; do not introduce local-model-backing optimization here.

---

## Required fail-closed controls

At minimum cover:

1. wrong accepted predecessor identity;
2. changed GGUF member hash;
3. changed runtime/binary hash;
4. stale/wrong GPU UUID/BDF/topology;
5. unresolved `greedy` sampler name falsely accepted;
6. `top_k>1` falsely accepted as true greedy;
7. missing sampler-chain evidence;
8. request mutation after reference freeze;
9. wrong prompt/token IDs;
10. candidate topology differing from frozen authority;
11. stale process/session contamination after restart;
12. output mismatch suppressed/rewritten;
13. authored PASS contradicting retained raw output;
14. server `n_probs` row treated as a trusted decision-row oracle despite the accepted R8-C limitation;
15. accepted R8-B/R8-C evidence mutation.

---

## Terminal classifications

Use exactly one:

### `R8D_QWEN38_TRUE_GREEDY_NVIDIA_RPC_QUALIFICATION_PASS`

Only if the sampler contract is prospectively and mechanically proven true-greedy on both arms, the frozen reference is deterministic, all four candidate cases are exact-token/stop identical to the reference, restart/repeatability passes, intended residency/device/RPC participation is proven, accounting is truthful, and all required controls pass.

### `R8D_QWEN38_TRUE_GREEDY_NVIDIA_RPC_QUALIFICATION_FAIL`

Use when a valid frozen candidate reaches correctness-bearing execution and any required exact-token, stop, placement, restart, sampler-contract, or accounting invariant fails.

Do not repair/tune and rerun under the same authority to convert a valid FAIL into PASS.

### `R8D_EVIDENCE_BLOCKED`

Use only before valid distributed correctness output when a concrete prerequisite cannot be established, including inability to prove the true-greedy sampler contract, missing/corrupt frozen bytes, non-deterministic reference, or insufficient trustworthy placement/accounting evidence.

A generated candidate correctness mismatch may not be relabeled BLOCKED.

---

## Explicit non-goals

Do not in R8-D:

- change llama.cpp revision to seek a PASS;
- repair llama.cpp;
- weaken exact-token equality after observing residual true-greedy divergence;
- reinterpret R8-B or R8-C historical terminals;
- introduce AMD/Vulkan/mixed-vendor execution;
- integrate Qwen into InferSwarm ordinary serving;
- change the generic planner;
- change quantization/model representation;
- add MTP/speculative or multimodal execution;
- implement node-local model-backing/source-policy optimization;
- claim production readiness.

---

## Validation and review gate

Before handoff require:

1. canonical bootstrap + doctor green;
2. focused R8-D authority/sampler/reducer/negative-control tests green;
3. accepted R8-A/R8-B/R8-C manifests byte-preserved;
4. full CPU suite green via `scripts/run_full_cpu_suite.py`;
5. CI planner/workflow parity green if touched;
6. project-status synchronization/check green;
7. finalizer fixed point + `--check` green;
8. hosted CI SUCCESS on exact final PR head;
9. two independent exact-head adversarial reviews:
   - sampler/correctness/reference/RPC semantics;
   - provenance/model identity/residency/accounting/non-claims;
10. resolve every P0/P1 and post final dispositions on the PR;
11. leave exactly one PR OPEN / UNMERGED for maintainer review.

---

## Required final report

Report at minimum:

1. starting/final main and PR heads;
2. accepted R8-C merge/terminal identity;
3. exact llama.cpp source/build/binary authority;
4. exact UD-IQ1_S split identities;
5. fresh host/GPU/NIC/runtime inventory identities;
6. exact frozen request JSON;
7. mechanical proof `top_k=1` truly constrains support to one token before the terminal draw;
8. sampler-chain/log evidence for reference and candidate;
9. old-invalid-`greedy` negative control result;
10. exact four prompt/token fixtures;
11. reference repeatability and frozen outputs/digests;
12. candidate outputs/digests;
13. per-case exact token/stop equality and first divergence where applicable;
14. >2K and 4097-token results;
15. candidate repeatability;
16. restart identity/reproof/result;
17. per-resource weight/PLE/KV/recurrent/workspace/staging accounting;
18. backing/mmap/RSS accounting;
19. fallback/wrong-device/source/state-movement findings;
20. required negative-control results;
21. descriptive startup/prefill/decode/network observations;
22. exact terminal;
23. evidence/manifest identities;
24. focused/full CPU test results + identity digest;
25. finalizer/status/CI parity;
26. exact-head hosted CI result;
27. both exact-head review dispositions;
28. confirmation R8-B/R8-C evidence remained byte-unchanged;
29. PR URL/head and OPEN / UNMERGED state.

Stop for maintainer review. Do not create the local-model-backing successor or any methodology-change successor in the same session.

**Comment by ezutfen** [comment]
## Completion disposition

Completed via PR #197, merged as `f142a0d9b693f999685960c641b2a8fe362c4e1e` from reviewed head `ca7e159929190a04cabd059042def1b6bd4c2467`.

Authoritative campaign: `issue195-r8d-true-greedy-requal-v2`.

Accepted terminal: `R8D_QWEN38_TRUE_GREEDY_NVIDIA_RPC_QUALIFICATION_FAIL`.

Accepted result: true-greedy execution materially improved agreement (1024/3072 exact) but residual deterministic divergence remains at case-256 position 5 and case-4096 immediate EOS after full 4097-token prefill. Group A/B qualification invariants passed; the failure is correctness-only. Future work is split into bounded residual-divergence characterization and generic model-backing/source-policy work rather than another R8-D qualification rerun. Closing this ticket as completed.
